### ISSUE TITLE
Using `await wait(...)` hangs jest test

### ISSUE THREAD

**Issue by DmacLoyaltyOne** [testing-library/react-testing-library#242]
- `react-testing-library` version: 5.3.0
- `jest` version: 23.6.0
- `react` version: 16.7 alpha 2
- `node` version: 11.0.0
- `npm` (or `yarn`) version: 6.4.1

### Relevant code or config:

```js

const MockComponent = ({ callback }) => {
  const ref = React.useRef(null)

  return (
    <div ref={ref}>
      <h1>Some Header</h1>

      <div>
        <h6>Some Child Header</h6>
      </div>
    </div>
  )
}

describe('GoodTest', async () => {
  test('callback should be called when clicking on document', async () => {

    const callback = jest.fn()

    const { getByText, unmount } = render(<MockComponent callback={callback} />)

    await waitForElement(() => getByText(/Some Header/))

   // I never see 'good' in the console
   // the test never completes
    await wait(() => {
      expect(1).toEqual(1)
      console.log('good')
    })

  })
})
```

### What you did:

I'm trying to test a component with a react hook that attaches a click listener on the document, but the listener is not applied by the time I am accessing it with the 'waitForElement' ...
So now, I'm using wait() to wait for something to appear, however, the test is never completing. 

### What happened:

My computer fan revs up like crazy, making me think it is stuck in an infinite loop.

### Reproduction:

I don't have a repo up yet, but I can set one up if that will help.

Something important to note was that I ran a single test with Jest doing `jest "GoodTest"` 
I am testing now to see if I get the same behaviour by running all tests.

**Comment by kentcdodds** [comment]
Hi @DmacLoyaltyOne,

Please make a reproduction. Thanks.

**Comment by DmacLoyaltyOne** [comment]
@kentcdodds Thanks for the quick response

https://github.com/DmacLoyaltyOne/ReactTestingLibrary-issue242

I simplified the wait() a bit to show that it keeps running no matter what I'm expecting inside.

In reality, I'd like to wait() until the map.mousedown is a function, meaning the eventhandler has been attached.

Seeing as you just did a series on Egghead testing react hooks, it would be interesting to see your take as to what is going wrong here.

**Comment by kentcdodds** [comment]
This probably has something to do with an issue that I saw with `useEffect`. I haven't investigated it, but I'm pretty sure that there's a bug in jsdom with `useEffect` where React's scheduler schedules it for later, but that later never comes for some reason... I haven't dug into it and don't have time to now. I would love to know what's going on though if you're interested in doing some [digging](https://kcd.im/dig).

What I recommend doing instead is using `flushEffects` and keeping things synchronous. I actually talked about this a bit in my newsletter today if that interests you: https://kcd.im/news

I'm going to go ahead and leave this open because I think this would be a good problem to figure out, but I'm fairly confident that the solution will not be in this project, but in jsdom (note: the reason I think it's a jsdom issue is because the problem does not surface when run in the browser).

**Comment by DmacLoyaltyOne** [comment]
@kentcdodds Thanks so much!
I appreciate your detailed answer and for taking a look at this so quickly. 🙏 
I don't have too much time to do spelunking at the moment, but I'll see what I can do, and I'll give the `flushEffects` a shot!

**Comment by DmacLoyaltyOne** [comment]
I may have found another issue. 
The flushEffects makes the code work correctly, but I notice the following code is not acting as expected.

```
fireEvent.click(document.body)
```
I would expect this to trigger the event listener that was added onto the document, but it does not seem to be triggering the callback.

If I intercept the event listener like this:
```
const map: any = {}

    document.addEventListener = jest.fn((event, cb) => {
      map[event] = cb
    })
```

and then trigger it manually:

```
map.mousedown({
      target: document.body
    })
```

Then it seems to work.

Should I open a separate issue for this, or is this just a limitation of the library?

**Comment by DmacLoyaltyOne** [comment]
Sure I'll give it a shot and see what I can come up with, and keep you updated.

**Comment by MiroslavPetrik** [comment]
I'm having the same issue. My flow is
1. Render component
2. `fireEvent.click` (click handler calls `setState`)
3. `queryByText` or `container.querySelector`

step 3 never finishes. For `queryByText` the test eventually crashes node with

```
debug> < FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
```
Because of some String allocations in replace or substr methods. Don't know if some of the query functions could cause this...

And the `container.querySelector` resolves to `null`.

When I call `debug()` after step 2. the printed pretty DOM contains the text, and elements which I'm trying to query (the structure is as expected). So it seems strange that the query calls don't work.

To be more specific, my rendered component is a Dropdown from `ant design` library connected to a Context. IDK if a demo setup would help, as the test itself is trivial - resembling hello world scenario...

**Comment by kentcdodds** [comment]
> IDK if a demo setup would help

A demo will always help. Please provide one.

**Comment by alexkrolick** [comment]
I can repro this with `waitForElement` and `wait`

**Comment by alexkrolick** [comment]
Possibly related to https://github.com/facebook/jest/issues/7339

**Comment by alexkrolick** [comment]
My issue looks like an infinite loop in a component, which might be triggering the "sync tests never timeout" issue (facebook/jest#6947).

**Comment by kentcdodds** [comment]
Going to close this one for inactivity and I don't think that it's a bug in react-testing-library anyway.

**Comment by benmonro** [comment]
@kentcdodds I think I may be seeing this... Has anyone happened to see it since this was closed?

**Comment by kentcdodds** [comment]
Not me

**Comment by ayushrajLT** [comment]
I just had this issue, might have to do with unfinished tests, especially if you have conditionals in your code and you didn't handle else case. not sure!
